chore: 🤖 Apply BestieTemplate v0.7.2

abelsiqueira opened the pull request:
Applied the template and made some manual changes, such as changing the
docs names, fixing the README and removing some older workflows.

✅ Closes: #143

abelsiqueira commented:
I suggest dropping Cirrus CI @tmigot, what do you think?

tmigot commented:
> I suggest dropping Cirrus CI @tmigot, what do you think?

I personally agree. What do you think @amontoison ?

amontoison commented:
@tmigot I opened #166 to fix CirrusCI.

abelsiqueira commented:
Thanks, @amontoison. I still think we should delete Cirrus CI testing from this package. I don't think the main use case of this package requires testing on FreeBSD or a specific Mac that GH doesn't already provide. This is essentially an interface, so the underlying packages should be more thoroughly tested, but this doesn't need to.
@tmigot, let me know if you still agree, I'll open an issue then. I will rebase this.

amontoison commented:
The issue was that MacOS silicon was not supported by GHA until a very recent update.

I think that Cirrus is relevant for packages that depend on Artifact (JLL packages) but we can remove it otherwise.

amontoison commented:
> > The issue was that MacOS silicon was not supported by GHA until a very recent update.
> > I think that Cirrus is relevant for packages that depend on Artifact (JLL packages) but we can remove it otherwise.
> 
> Thanks! So, still good to me to remove cirrus for this package.

Yes, we just need to update GHA to test on two architectures (x64 and aarch64). macos-13 is the last release of Mac that supports this architecture. macos-lastest is only working on aarch64 now.

abelsiqueira commented:
I've updated the PR with the latest version of the template. I've removed CirrusCI. If we apply the template to other packages in JSO that use CirrusCI, we can update the template with the latest file.

As for the extra tests, I also don't see the benefit of adding them now, in this package.

The Lint failures are due to broken links. After merging, most or all should be fixed.

dpo commented:
> I've removed CirrusCI.

Why?

abelsiqueira commented:
> Thanks, @amontoison. I still think we should delete Cirrus CI testing from this package. I don't think the main use case of this package requires testing on FreeBSD or a specific Mac that GH doesn't already provide. This is essentially an interface, so the underlying packages should be more thoroughly tested, but this doesn't need to. @tmigot, let me know if you still agree, I'll open an issue then. I will rebase this.

@dpo, ☝️

dpo commented:
> The issue was that MacOS silicon was not supported by GHA until a very recent update.
> 
> I think that Cirrus is relevant for packages that depend on Artifact (JLL packages) but we can remove it otherwise.

With that logic, there's no need to test on Windows or macOS either. I disagree. Do the FreeBSD tests not pass? The more platforms we cover, the more robust our code.

abelsiqueira commented:
I think the cost-benefit of adding FreeBSD is not good enough, but it is for Mac and Windows. That being said, if it makes it easier to adopt COPIERTemplate, I'm glad to reintroduce it. The file in the template is the following: https://github.com/abelsiqueira/COPIERTemplate.jl/blob/main/template/%7B%25%20if%20UseCirrusCI%20%25%7D.cirrus.yml%7B%25%20endif%20%25%7D.jinja
I've added the file there because JSO uses Cirrus CI, so can you update that file to something that makes sense for JSO?
I can also remove from there so it is used directly here.

My arguments for not having FreeBSD in this package are:

- 1% of Julia Users are on FreeBSD, and who knows how many of those will be using JSOSuite.
- Most likely we'll pick those errors on earlier packages, and if we do find an error with FreeBSD that no other package upstream had, then we can't properly test it.
- it is a separate build script to maintain.
- the building hours are shared among all JSO packages, so this is just wasting those.

abelsiqueira commented:
@dpo @tmigot can we have a conclusion on this?

dpo commented:
I just find it arbitrary to remove Cirrus from here while we use it everywhere else. I've found issues with code before because errors were triggered on a Cirrus VM. If we can cover another OS, why not? Predicting the effects of a change in one of the dependent packages is very hard, so the more coverage, the better, in my opinion.

abelsiqueira commented:
@dpo can you fit in your schedule maintaining the [template's Cirrus file](https://github.com/abelsiqueira/COPIERTemplate.jl/blob/main/template/%7B%25%20if%20UseCirrusCI%20%25%7D.cirrus.yml%7B%25%20endif%20%25%7D.jinja?rgh-link-date=2024-06-11T18%3A54%3A48Z)?

dpo commented:
Sure. there's really nothing to it.

abelsiqueira commented:
Great, I'll update this then

tmigot commented on .github/ISSUE_TEMPLATE/10-bug-report.yml:
```diff
@@ -0,0 +1,61 @@
+        Please, before submitting, make sure that:
+
+        - There is not an [existing issue](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/issues) with the same question
+        - You have read the [contributing guide](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/docs/src/90-contributing.md/)
```
Should the link point to the documentation in the end?

abelsiqueira replied:
Will update the template.

tmigot commented on .github/ISSUE_TEMPLATE/10-bug-report.yml:
```diff
@@ -0,0 +1,61 @@
+        - You have read the [contributing guide](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/docs/src/90-contributing.md/)
+        - You are following the [code of conduct](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/CODE_OF_CONDUCT.md)
+
+        The form below should help you in filling out this issue.
```
I think we should more directly ask for a Minimal reproducible example. From what I've seen 75% of the first comment from a question is "Do you have a MWE?"

abelsiqueira replied:
Will update the template. There is a question for reproduction steps, I will add more info there.

tmigot commented on .github/ISSUE_TEMPLATE/20-feature-request.yml:
```diff
@@ -0,0 +1,37 @@
+name: "Feature Request"
+description: Suggest a new feature for the package
+body:
```
Miss the title/label sections here?

abelsiqueira replied:
The title for the bug had a prefix "[Bug] " to make the issue more evident as a bug. The others will have a default empty title. Same for the label.

tmigot commented on .github/ISSUE_TEMPLATE/20-feature-request.yml:
```diff
@@ -0,0 +1,37 @@
+  - type: markdown
+    attributes:
+      value: |
+        Thanks for taking the time to fill out this bug report!
```
```suggestion
        Thanks for taking the time to fill out this feature request report!
```

tmigot commented on .github/ISSUE_TEMPLATE/20-feature-request.yml:
```diff
@@ -0,0 +1,37 @@
+        - You have read the [contributing guide](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/docs/src/90-contributing.md/)
+        - You are following the [code of conduct](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/CODE_OF_CONDUCT.md)
+
+        The form below should help you in filling out this issue.
```
We could also suggest to explain ways to test the new request?

tmigot commented on .github/ISSUE_TEMPLATE/30-usage.yml:
```diff
@@ -0,0 +1,24 @@
+  - type: markdown
+    attributes:
+      value: |
+        Thanks for taking the time to fill out this bug report!
```
```suggestion
        Thanks for taking the time to fill out this documentation report!
```

tmigot commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+- [ ] Tests are passing
+- [ ] Lint workflow is passing
+- [ ] Docs were updated and workflow is passing
+<!-- - [ ] [CHANGELOG.md](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/CHANGELOG.md) was updated -->
```
remove?

tmigot commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,7 @@
+  - package-ecosystem: "github-actions"
+    directory: "/" # Location of package manifests
+    schedule:
+      interval: "weekly"
```
Is this some kind of CompatHelper?

tmigot commented on README.md:
```diff
@@ -1,51 +1,34 @@
-[release-url]: https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/releases
-
-One stop solutions for all things optimization.
+# JSOSuite
```
It is probably worth adding somewhere adding that we use the COPIERTemplate in case somebody tries to update CI and stuff. what do you think?

tmigot commented on README.md:
```diff
@@ -1,51 +1,34 @@
 
-If you use JSOSuite.jl in your work, please cite using the format given in [CITATION.cff](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/blob/main/CITATION.cff).
-
-## Installation
```
Any reason to remove this? and the basic example?

tmigot commented on docs/src/index.md:
```diff
@@ -1,113 +1,20 @@
-# JSOSuite.jl
```
I think some content has been lost here

tmigot reviewed (changes requested):
Hi @abelsiqueira ! Sorry for the delay, but this is a massive PR^^. Are the modification to the code generated or some of them manually modified?
Feel free to let me know if you prefer that I open issues directly to COPIERTemplate

abelsiqueira commented:
I've updated to BestieTemplate (it has been renamed) version 0.7.1.
The Cirrus CI is failing right now, so I don't know how you want to proceed @dpo

amontoison commented:
CirrusCI changed the rules recently and we only have a (very) limited number of credits now.
We should drop it after this update

abelsiqueira commented:
If we're dropping it, it should be here so we can squash merge this, and because it is one of the question of the template and if we drop it we simplify the application more.

abelsiqueira commented:
@amontoison can you explain more on the changes or the limitation? @dpo had arguments to maintain, but if the limitation is too severe I imagine it might affect using Cirrus CI on other packages as well.
Here's the original message:

> I just find it arbitrary to remove Cirrus from here while we use it everywhere else. I've found issues with code before because errors were triggered on a Cirrus VM. If we can cover another OS, why not? Predicting the effects of a change in one of the dependent packages is very hard, so the more coverage, the better, in my opinion.

tmigot commented on README.md:
```diff
@@ -1,51 +1,34 @@
+[![Docs workflow Status](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/actions/workflows/Docs.yml/badge.svg?branch=main)](https://github.com/JuliaSmoothOptimizers/JSOSuite.jl/actions/workflows/Docs.yml?query=branch%3Amain)
+[![Build Status](https://api.cirrus-ci.com/github/JuliaSmoothOptimizers/JSOSuite.jl.svg)](https://cirrus-ci.com/github/JuliaSmoothOptimizers/JSOSuite.jl)
+[![Coverage](https://codecov.io/gh/JuliaSmoothOptimizers/JSOSuite.jl/branch/main/graph/badge.svg)](https://codecov.io/gh/JuliaSmoothOptimizers/JSOSuite.jl)
+[![DOI](https://zenodo.org/badge/DOI/FIXME)](https://doi.org/FIXME)
```
```suggestion
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.12588054.svg)](https://doi.org/10.5281/zenodo.12588054)
```
https://zenodo.org/records/12588054

tmigot reviewed (approved):
Thanks @abelsiqueira ! It looks good to me overall. I think it is a good place to have some experience on BestieTemplate.jl + it is quite easy to update with the template.

I would say just two things before merge:
- fix the DOI (see my comment)
- clarify cirrus

Thanks!!

dpo commented:
CirrusCI gives us 10,000 minutes (= 7 x 24 hours) per month for Linux VMs and 5,000 minutes for macOS VMs. Is that such a limitation??? I'm a bit bummed by these discussions.

The failure in this PR is unrelated to time limits.

dpo commented on .cirrus.yml:
```diff
@@ -2,14 +2,14 @@ task:
     - name: FreeBSD
       freebsd_instance:
-        image_family: freebsd-13-3
+        image_family: freebsd-13-2
```
```suggestion
        image_family: freebsd-13-3
```

dpo commented on .cirrus.yml:
```diff
@@ -2,14 +2,14 @@ task:
     - name: MacOS M1
       macos_instance:
-        image: ghcr.io/cirruslabs/macos-ventura-base:latest
+        image: ghcr.io/cirruslabs/macos-monterey-base:latest
```
```suggestion
        image: ghcr.io/cirruslabs/macos-sonoma-base:latest
```

abelsiqueira commented:
Can you submit a RP to update https://github.com/abelsiqueira/BestieTemplate.jl/? Otherwise it defeats the purpose of having Cirrus it on the template.

abelsiqueira commented:
I've updated to 0.7.2 which includes @dpo's changes for Cirrus, and I merged @tmigot's suggestion, so it is done and I will merge now. Thanks everyone for their time and input.
